File: lib/shared/models/api_response.dart
class ApiResponse<T> {
  final bool success;
  final String message;
  final T? data;
  final Map<String, dynamic>? meta;
  final List<String> errors;

  const ApiResponse({
    required this.success,
    required this.message,
    this.data,
    this.meta,
    this.errors = const [],
  });

  factory ApiResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Object?) fromJsonT,
  ) =>
      ApiResponse<T>(
        success: json['success'] as bool? ?? false,
        message: json['message'] as String? ?? '',
        data: json['data'] != null ? fromJsonT(json['data']) : null,
        meta: json['meta'] as Map<String, dynamic>?,
        errors: (json['errors'] as List<dynamic>?)
                ?.map((e) => e.toString())
                .toList() ??
            [],
      );
}

class PaginationMeta {
  final int page;
  final int limit;
  final int total;
  final int totalPages;
  final bool hasNext;
  final bool hasPrev;

  const PaginationMeta({
    required this.page,
    required this.limit,
    required this.total,
    required this.totalPages,
    required this.hasNext,
    required this.hasPrev,
  });

  factory PaginationMeta.fromJson(Map<String, dynamic> json) => PaginationMeta(
        page: json['page'] as int? ?? 1,
        limit: json['limit'] as int? ?? 20,
        total: json['total'] as int? ?? 0,
        totalPages: json['totalPages'] as int? ?? 1,
        hasNext: json['hasNext'] as bool? ?? false,
        hasPrev: json['hasPrev'] as bool? ?? false,
      );
}

File: backend/src/utils/apiResponse.js
const sendResponse = (res, { status, success, message, data, meta }) => {
  const response = { success, message };
  if (data !== undefined) response.data = data;
  if (meta !== undefined) response.meta = meta;
  return res.status(status).json(response);
};

const sendSuccess = (res, message, data, meta, status = 200) =>
  sendResponse(res, { status, success: true, message, data, meta });

const sendCreated = (res, message, data) =>
  sendResponse(res, { status: 201, success: true, message, data });

const sendError = (res, message, status = 500, errors) => {
  const response = { success: false, message };
  if (errors) response.errors = errors;
  return res.status(status).json(response);
};

module.exports = { sendSuccess, sendCreated, sendError };
